once/_iterator_wrappers.py

```diff
@@ -24,116 +25,156 @@ def __init__(self) -> None:
-class AsyncGeneratorWrapper:
+
+class _IteratorAction(enum.Enum):
+    # Generating the next value from the underlying iterator
+    GENERATING = 1
+    # Yield an already computed value
+    YIELDING = 2
+    # Waiting for the underlying iterator, already triggered from another call.
+    WAITING = 3
+    # We can return, we are done!
+    RETURNING = 4
+
+
+class _GeneratorWrapperBase:
+    """Base class for generator wrapper.
+
+    Even though the class stores a result, all of the methods separately take a result input.
+    Why is that? Great question.
+    During a call to yield_results, we should grab a reference to the existing self.result, and
+    only operate on that by passing it into here. In the event of an Exception, a new execution is
+    kicked off to retry, but existing iterators from yield_results will continue as if that never
+    happened, and still raise an Exception. This will avoid mixing results from different
+    iterators.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.callable: collections.abc.Callable | None = functools.partial(func, *args, **kwargs)
+        self.generator = self.callable()
+        self.result = IteratorResults()
+        self.generating = False
+
+    def compute_next_action(
+        self, result: IteratorResults, i: int
+    ) -> typing.Tuple[_IteratorAction, typing.Any]:
+        """Must be called with lock."""
+        if i == len(result.items):
+            if result.finished:
+                if result.exception:
+                    raise result.exception
+                return _IteratorAction.RETURNING, None
+            if self.generating:
+                return _IteratorAction.WAITING, None
+            else:
+                # If all of these functions are called with locks, we will never have more than one
+                # caller have GENERATING at any time.
+                self.generating = True
+                return _IteratorAction.GENERATING, None
+        else:
+            return _IteratorAction.YIELDING, result.items[i]
+
+    def successful_completion(self, result: IteratorResults):
+        """Must be called with lock."""
+        result.finished = True
+        self.generating = False
+        self.generator = None  # Allow this to be GCed.
+        self.callable = None  # Allow this to be GCed.
+
+    def record_item(self, result: IteratorResults, item: typing.Any):
+        self.generating = False
+        result.items.append(item)
+
+    def record_exception(self, result: IteratorResults, exception: Exception):
+        """Must be called with lock."""
+        result.finished = True
+        # We need to keep track of the exception so that we can raise it in the same
+        # position every time the iterator is called.
+        result.exception = exception
+        self.generating = False
+        assert self.callable is not None
+        self.generator = self.callable()  # Reset the iterator for the next call.
```

Comment:
As discussed earlier today, I think by default we should not reset the iterator and just close out here. We can add this in when we support configuration

Reply:
it is now configurable!